src/lib/optimizer/abstract_syntax_tree/abstract_ast_node.hpp

```diff
@@ -226,6 +238,10 @@ class AbstractASTNode : public std::enable_shared_from_this<AbstractASTNode> {
   std::vector<std::string> get_verbose_column_names() const;
   // @}
 
+  // Returns a deep copy of this subtree. If your subclass holds any shared_ptr or other members whose copy
+  // constructor does not give you a real copy, you have to override this
+  virtual std::shared_ptr<AbstractASTNode> clone_subtree() const = 0;
```

Comment:
There is no condition on the "have to override", as you don't provide a default implementation. Also, we use "recreate" in the Operators and "clone_subtree" here for the semantics. Imo "recreate" is cool.

Reply:
I feel like the semantics are different. `clone` is a 1:1 deep copy while `recreate` also replaces the placeholders.